Add types using JSDoc to the `graphlib` file

aloisklink opened the pull request:
Similar to https://github.com/tbo47/dagre-es/pull/45, I've gone through the `src/graphlib/graph.js` file and added proper types to the functions whenever TypeScript couldn't automatically figure out the types.

I've also added type-checking of the `*.test.js` files to ensure that at least the types match what the unit tests expect!

### Stuff I didn't do and why

There's also an MIT-licensed [`@types/graphlib`][1] package that we can copy code from, but I didn't do that since I noticed issues with it that were causing our unit tests to fail type-checking (e.g. not allowing `number` params for lots of functions, even though the test code uses it a lot). We could copy it over later if we want to have better descriptions for the graphlib functions.

[1]: http://npmjs.com/package/@types/graphlib

I also tried to apply 8740b83207d639ff9bfd61a5117e381db8db474a from https://github.com/tbo47/dagre-es/pull/45 and unfortunately it seemed to cause errors in other parts of the code, since they're still not typed properly. But all the graphlib types looked correct to me.

aloisklink committed:
test: type `var conflicts` vars in `bk.test.js`

This fixes any potential type errors when calling the `addConflict`
function in our unit tests, after the changes from
dbe53cb (fix: Prototype Pollution (#54), 2025-10-21).

Fixes: dbe53cb81afd318d9d800bb21167374df262b8df

aloisklink committed:
test: import missing `beforeEach` and `afterEach`

Vitest adds these automatically, so it didn't cause any issues when
running the tests, but it did cause issues with TypeScript.

aloisklink committed:
test: type `var g` with `Graph`

This lets TypeScript check that the functions we call on this type are
valid.

aloisklink committed:
build: change `tsconfig.json` to type-check tests

Update the `tsconfig.json` file to also type-check the tests, checking
for any potential mistakes we make in TypeScript.

The new `tsconfig.build.json` file is then used for actually building
the types.

aloisklink committed:
feat(types): improve `Graph` documented types

I've manually gone through the JSDoc references within the
`src/graphlib/graph.js` file and added JSDoc comments when TypeScript
couldn't automatically find detect the types.

I've tried to keep the types as backwards compatible as possible.

There's also an MIT-licensed [`@types/graphlib`][1] package that we can
copy code from, but I didn't do that since I noticed issues with it that
were causing our unit tests to fail type-checking (e.g. not allowing
`number`s as params for lots of functions, even though the test code
uses it a lot).

[1]: http://npmjs.com/package/@types/graphlib

aloisklink committed:
fix(types): improve horizontalCompaction type

tbo47 commented:
Do you want me to release a new version for it? Or do you want to test more in mermaid first?

aloisklink commented:
I've tested it with Mermaid and it looks okay to me!

It even caught some bugs in Mermaid, there was some code calling graphlib with the args in the wrong order, see:

https://github.com/mermaid-js/mermaid/blob/b1fe4ffe97ba74f49c5a3d3058508a3bffd8e0c6/packages/mermaid/src/rendering-util/createGraph.ts#L124-L125

Up to you if you want to make a new release or not :shrug:.

This PR is just a types change, so it shouldn't affect any actual code, but it could be worth releasing this with https://github.com/tbo47/dagre-es/pull/51.

It's a long weekend in Japan, so if I have free time, I might make another PR trying to add types to the rest of the the `src/graphlib` folder. But it's sometimes hard to find free time to work on open-source :sweat_smile:.

aloisklink committed:
docs: copy over graphlib documentation for Graph

Update the Graph documentation with a mixture of:

1. API documentation from the graphlib wiki:
   https://github.com/dagrejs/graphlib/wiki/API-Reference
2. Documentation from
   https://github.com/dagrejs/graphlib/blob/64b31ef3df0270c9bda85b45cb338481ce25618d/lib/graph.js

Both of these are from the graphlib MIT license, which we're
already complying with.

aloisklink commented:
I've just pushed another change. I've copied over the official documentation from graphlib with a mixture of:

1. API documentation from the graphlib wiki: https://github.com/dagrejs/graphlib/wiki/API-Reference
2. Documentation from https://github.com/dagrejs/graphlib/blob/64b31ef3df0270c9bda85b45cb338481ce25618d/lib/graph.js

Both of these are from the graphlib MIT license, which we're already complying with.
